scrapy_google_sheets_exporter/gsheets_exporter.py

```diff
@@ -0,0 +1,79 @@
+import csv
+import logging
+from io import StringIO
+
+import gspread
+from scrapy.exceptions import NotConfigured
+from scrapy.extensions.feedexport import BlockingFeedStorage, build_storage
+
+logger = logging.getLogger(__name__)
+
+
+class GoogleSheetsFeedStorage(BlockingFeedStorage):
+    def __init__(self, uri, credentials, *, feed_options=None):
+        if credentials:
+            self.gc = gspread.service_account_from_dict(credentials)
+        else:
+            raise NotConfigured(
+                "Must specify GOOGLE_CREDENTIALS (dict) in the spider settings."
+            )
+
+        self.spreadsheet_key, self.sheet_name = self.parse_gsheets_uri(uri)
+        self.spreadsheet = self.gc.open_by_key(self.spreadsheet_key)
+        self.sheet = self.spreadsheet.worksheet(self.sheet_name)
+        self.feed_options = feed_options or {}
+        self.append_mode = self.feed_options.get("append_mode", False)
+        self.fields = self.feed_options.get("fields", [])
+        self.format = self.feed_options.get("format", "csv")
+
+        if self.format != "csv":
+            raise NotConfigured(
+                "This feed exporter only supports csv format. "
+                f"Please update the FEEDS settings by replacing {self.format} format."
+            )
+
+    @classmethod
+    def from_crawler(cls, crawler, uri, *, feed_options=None):
+        return build_storage(
+            cls,
+            uri,
+            credentials=crawler.settings["GOOGLE_CREDENTIALS"],
+            feed_options=feed_options,
+        )
+
+    def _store_in_thread(self, file):
+        file.seek(0)
+        csv_text = file.read().decode("utf-8")
+        csv_reader = csv.reader(StringIO(csv_text))
+        csv_data = list(csv_reader)
+
+        items = [dict(zip(csv_data[0], item)) for item in csv_data[1:]]
+
+        header = self.fields or csv_data[0]
+        if not self.append_mode:
+            self.sheet.clear()
+            self.sheet.append_row(header)
+
+        else:
+            if self.sheet.row_values(1):
+                header = self.sheet.row_values(1)
+                logger.warning(
+                    "FEED option 'append_mode' was set to True. "
+                    f"The following fields will be exported: {header}."
+                )
+            else:
+                self.sheet.append_row(header)
+
+        for item in items:
+            row = [item.get(field) for field in header]
+            self.sheet.append_row(row)
```

Comment:
We should try batching this call if possible. Otherwise, it will quite slow if we have to make one HTTP call per row and we also may run into quota / limits. 

You can use [append_rows](https://docs.gspread.org/en/v5.7.1/api/models/worksheet.html#gspread.worksheet.Worksheet.append_rows) instead for that. Also, combining it with batching to avoid pushing too many rows at the same time.

Reply:
@VMRuiz I changed to append_rows as you suggested. 
@ftadao what do you think should be our default batch size (byte wise)?

Reply:
@VMRuiz @ftadao @Gallaecio my current solution appends all rows at once without batching it. I tried using the `batch_item_count` feed option (same as https://docs.scrapy.org/en/latest/topics/feed-exports.html#std-setting-FEED_EXPORT_BATCH_ITEM_COUNT) but this requires multiple output files, which is not our use case here since we'll be exporting to only 1 google sheet and we don't have a way to provide the `%batch_id` identifier in the URI.

How do you suggest I handle the batching? should we have a custom feed option like `batch_size` that the user should set or should we let the exporter handle this in another way (without the user having to set anything)?

Reply:
Hello @liviasalgueiro , 

You are right, adding `FEED_EXPORT_BATCH_ITEM_COUNT` would require to pass `%(batch_id)d` as part of the URL, and that doesn't looks like a good solution. 

Instead, I suggest defining an exporter parameter like `batch_size` that can be configure by using the `item_export_kwargs` field. 

If not value is provided then we use a default batching value. I don't know right now what would be a good value for such default. From google docs: `Google recommends a 2-MB maximum payload`

Reply:
Also, @Gallaecio @ftadao, this brings back the idea of implementing Stream Exporters, which aren't based on underlaying files, instead of using Feed Exporters

Reply:
@VMRuiz I added batching as suggested, please let me know what you think. Thank you!

Reply:
It's still not the ideal solution as we're still saving everything to a local file and then sending it in batches instead of sending batches directly. As @VMRuiz noted this is another point towards the creation of a Stream Exporter.

We can release this exporter as it is working right now, and plan for the Stream Exporter development. Code LGTM.